File: PELT_Code/PELT.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cstring>
#include <memory>
#include "multiple_processes_regime.h"
using namespace std;

void calculate_F_tau_star_and_tau_1(vector< double > &, vector< unsigned long int > &, double &, unsigned long int &, unsigned long int, mult_process&);
void set_R_tau_star_plus_1(vector< unsigned long int > &, vector< unsigned long int > &, vector< double > &, unsigned long int, const double &, mult_process&);
void write_cp_n_to_file(vector< unsigned long int > &, string, const unsigned long int &);
void print_asterisk_if_needed(unsigned long int, unsigned long int, unsigned long int &, unsigned long int);

int main(int argc, char * argv[]) {
    string data_file = ""; // name of the data file
    string penalty = "BIC"; // can be BIC or AIC
    unsigned long int end_time = 1; // assuming that the process starts at time 0 (with the first transition happening at time 1 for a Markov chain), when does it end?
    unsigned long int diff = 1; // squash multiple observations into a single time
    double beta = 0; // beta will be assigned in mult_process
    double K = 0;
    
    if(argc > 1) {
        ifstream InputParameters(argv[1], ios::in);
        InputParameters >> data_file;
        InputParameters >> penalty;
        InputParameters >> end_time;
        InputParameters >> diff;
    }
    
    std::unique_ptr<mult_process> pm = std::unique_ptr<mult_process>(new mult_process(data_file, penalty, end_time, diff, beta));
    end_time /= diff;
  
    vector< double > F(end_time + 2, 0);
    vector< vector< unsigned long int > > R(end_time + 3);
    vector< vector< unsigned long int > > cp(end_time + 2);
  
    F[0] = -beta;
    R[1].push_back(0);
    unsigned long int number_of_asterisks_already_printed = 0;

    for (unsigned long int tau_star = 1; tau_star <= end_time + 1; tau_star++) {
        unsigned long int tau_1;
        calculate_F_tau_star_and_tau_1(F, R[tau_star], beta, tau_1, tau_star, *pm);
        cp[tau_star] = cp[tau_1];
        cp[tau_star].push_back(tau_1);
        set_R_tau_star_plus_1(R[tau_star + 1], R[tau_star], F, tau_star, K, *pm);
        print_asterisk_if_needed(tau_star, end_time, number_of_asterisks_already_printed, 10);
    }
    
    cout << "\n";
    
    data_file.erase(data_file.end() - 15, data_file.end());
    string cp_n_Filename = data_file + "_cp_n.txt";
    write_cp_n_to_file(cp.back(), cp_n_Filename, diff);
    
    return 0;
}

void calculate_F_tau_star_and_tau_1(vector< double > & F, vector< unsigned long int > & R_tau_star, double & beta, unsigned long int & tau_1, unsigned long int tau_star, mult_process & pm) {
    tau_1 = R_tau_star[0];
    double min = F[tau_1] + pm.cost(tau_1, tau_star) + beta;
    for (unsigned int index = 1; index < R_tau_star.size(); index++) {
        unsigned long int tau = R_tau_star[index];
        double cost = F[tau] + pm.cost(tau, tau_star) + beta;
        if (cost < min) {
            min = cost;
            tau_1 = tau;
        }
    }
    F[tau_star] = min;
}

void set_R_tau_star_plus_1(vector< unsigned long int > & R_tau_star_plus_1, vector< unsigned long int > & R_tau_star, vector< double > & F, unsigned long int tau_star, const double & K, mult_process & pm) {
    for (unsigned int index = 0; index < R_tau_star.size(); index++) {
        if (F[R_tau_star[index]] + pm.cost(R_tau_star[index], tau_star) + K <= F[tau_star]) {
            R_tau_star_plus_1.push_back(R_tau_star[index]);
        }
    }
    R_tau_star_plus_1.push_back(tau_star);
}

void write_cp_n_to_file(vector< unsigned long int > & cp_n, string output_filename, const unsigned long int & diff) {
    ofstream OutputStream(output_filename, ios::out);
    OutputStream << setiosflags(ios::fixed);
    for (unsigned long int i = 1; i < cp_n.size(); i++) {
        OutputStream << cp_n[i] * diff << endl;
    }
    OutputStream.close();
}

void print_asterisk_if_needed(unsigned long int x, unsigned long int n, unsigned long int & number_already_printed, unsigned long int num_asterisks) {
    if (x > n * number_already_printed / num_asterisks) {
        cout << "*";
        cout.flush();
        number_already_printed++;
    }
}

File: PELT_Code/PELT.py
import numpy as np

class probability_model:
    def __init__(self, x, end_time, beta, likelihood_model):
        self.x = x
        self.end_time = end_time
        self.beta = beta
        self.likelihood_model = likelihood_model
    
    def cost(self, t0, t1):
        if self.likelihood_model == "gaussian":
            n = t1 - t0
            # check if there is only one observation in the interval
            if n == 1:
                # the likelihood for the observation will be 1, so log(likelihood) = 0
                return 0
            
            mu_hat = np.mean(self.x[t0:t1])
            sigma_squared_hat = np.std(self.x[t0:t1])
    
            if sigma_squared_hat < 0:
                raise ValueError("Gaussian variance estimated to be below 0")
    
            # check if the variance is 0
            if sigma_squared_hat < 0.0000001:
                # the likelihood for the observation will be 1, so log(likelihood) = 0
                return 0
    
            maximum_log_likelihood = -(np.log(2*np.pi) + np.log(sigma_squared_hat) + 1) * n / 2
            return -maximum_log_likelihood * 2.0
        else:
            raise ValueError("Model not implemented")


def calculate_F_tau_star_and_tau_1(F, R_tau_star, beta, tau_1, tau_star, pm):
    tau_1[0] = R_tau_star[0]
    min_cost = F[tau_1[0]] + pm.cost(tau_1[0], tau_star) + beta
    for tau in R_tau_star[1:]:
        cost = F[tau] + pm.cost(tau, tau_star) + beta
        if cost < min_cost:
            min_cost = cost
            tau_1[0] = tau
    F[tau_star] = min_cost


def set_R_tau_star_plus_1(R_tau_star_plus_1, R_tau_star, F, tau_star, K, pm):
    for tau in R_tau_star:
        if F[tau] + pm.cost(tau, tau_star) + K <= F[tau_star]:
            R_tau_star_plus_1.append(tau)
    R_tau_star_plus_1.append(tau_star)


#def print_asterisk_if_needed(x, n, number_already_printed, num_asterisks):
#    while x > n * number_already_printed[0] / num_asterisks:
#        print("*")
#        number_already_printed[0] += 1


def PELT_cp_n(x, end_time, likelihood_model = "gaussian", K = 0, penalty = "BIC"):
    if penalty == "BIC":
        beta = 2.0 * np.log(end_time + 1)
    elif penalty == "AIC":
        beta = 2.0 * 2.0
    
    pm = probability_model(x, end_time, beta, likelihood_model)

    F = np.zeros(end_time + 2)
    R = [[] for _ in range(end_time + 3)]
    cp = [[] for _ in range(end_time + 2)]

    F[0] = -beta
    R[1].append(0)
    number_of_asterisks_already_printed = [0]

    for tau_star in range(1, end_time + 2):
        tau_1 = [0]
        calculate_F_tau_star_and_tau_1(F, R[tau_star], beta, tau_1, tau_star, pm)
        cp[tau_star] = cp[tau_1[0]][:]
        cp[tau_star].append(tau_1[0])
        set_R_tau_star_plus_1(R[tau_star + 1], R[tau_star], F, tau_star, K, pm)
        #print_asterisk_if_needed(tau_star, end_time, number_of_asterisks_already_printed, 10)

    return cp